File: src/main/java/com/EmosewaPixel/pixellib/tiles/RecipeBasedTE.java
package com.EmosewaPixel.pixellib.tiles;

import com.EmosewaPixel.pixellib.recipes.SimpleMachineRecipe;
import com.EmosewaPixel.pixellib.recipes.SimpleRecipeList;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntityType;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class RecipeBasedTE extends AbstractRecipeBasedTE<SimpleMachineRecipe> {
    public RecipeBasedTE(TileEntityType type, SimpleRecipeList recipeList) {
        super(type, recipeList);
        setCurrentRecipe(SimpleMachineRecipe.EMPTY);
    }

    public SimpleMachineRecipe getRecipeByInput() {
        Stream<ItemStack> stacksStream = IntStream.range(0, getRecipeList().getMaxInputs()).mapToObj(i -> recipeInventory.getStackInSlot(i));

        if (stacksStream.anyMatch(ItemStack::isEmpty))
            return SimpleMachineRecipe.EMPTY;

        SimpleMachineRecipe chosenRecipe = getRecipeList().getRecipes().stream().filter(recipe -> recipe.isInputValid((ItemStack[]) stacksStream.toArray())).findFirst().get();

        if (chosenRecipe == null)
            return SimpleMachineRecipe.EMPTY;

        List<Integer> recipeIndexes = IntStream.range(0, getRecipeList().getMaxInputs())
                .map(i -> chosenRecipe.getIndexOfInput(recipeInventory.getStackInSlot(i))).boxed().collect(Collectors.toList());

        if (recipeIndexes.contains(-1))
            return SimpleMachineRecipe.EMPTY;

        return new SimpleMachineRecipe(
                IntStream.range(0, getRecipeList().getMaxInputs()).mapToObj(i -> new ItemStack(recipeInventory.getStackInSlot(i).getItem(), chosenRecipe.getCountOfInput(recipeIndexes.get(i)))).toArray(),
                (Float[]) recipeIndexes.stream().map(chosenRecipe::getConsumeChance).toArray(),
                chosenRecipe.getAllOutputs(),
                chosenRecipe.getOutputChances(),
                chosenRecipe.getTime());
    }
}

File: src/main/java/com/emosewapixel/pixellib/tiles/RecipeBasedTE.kt
package com.emosewapixel.pixellib.tiles

import com.emosewapixel.pixellib.recipes.SimpleMachineRecipe
import com.emosewapixel.pixellib.recipes.SimpleRecipeList
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntityType

open class RecipeBasedTE(type: TileEntityType<*>, recipeList: SimpleRecipeList) : AbstractRecipeBasedTE<SimpleMachineRecipe>(type, recipeList) {
    override var currentRecipe = SimpleMachineRecipe.EMPTY

    public override val recipeByInput: SimpleMachineRecipe
        get() {
            val stacksStream = (0 until recipeList.maxInputs).map { recipeInventory.getStackInSlot(it) }

            if (stacksStream.any { it.isEmpty })
                return SimpleMachineRecipe.EMPTY

            val chosenRecipe = recipeList.recipes.first { it.isInputValid(stacksStream.toTypedArray()) }
                    ?: return SimpleMachineRecipe.EMPTY

            val recipeIndexes = (0 until recipeList.maxInputs)
                    .map { chosenRecipe.getIndexOfInput(recipeInventory.getStackInSlot(it)) }

            return if (recipeIndexes.contains(-1)) SimpleMachineRecipe.EMPTY else SimpleMachineRecipe(
                    (0 until recipeList.maxInputs).map { ItemStack(recipeInventory.getStackInSlot(it).item, chosenRecipe.getCountOfInput(recipeIndexes[it])) }.toTypedArray(),
                    recipeIndexes.map { chosenRecipe.getConsumeChance(it) }.toTypedArray(),
                    chosenRecipe.outputs,
                    chosenRecipe.outputChances,
                    chosenRecipe.time)
        }
}